README.md
I2C LCD module powered by ACM1602  
=================================  
  



CHANGELOG.md
## 1.1 (12月 21,2018)

## 1.0.1 (12月 07, 2018)
  - Merge from the branch release-1.0  v1.0.0...HEAD
  - Be able to print Raspi Statuses on LCD 	modified:   acm1602.py 	new file:   yoko_scroll.py  v1.0.0...HEAD
  - Merge from the branch release-0.1  v1.0.0...HEAD
  - Create new release-1.0 branch  v1.0.0...HEAD

## 1.0 (12月 07,2018)
  - Merge from the branch release-0.1  v0.1.0...HEAD
  - Reconstruct 	new file:   .gitignore 	modified:   acm1602.py 	deleted:    tests/__pycache__/character_table.cpython-35.pyc  v0.1.0...HEAD
  - restrut  v0.1.0...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v0.1.0.  v0.1.0...HEAD
  - first commi  v0.1.0...HEAD

## 0.1.0 (12月 06,2018)




tests/char.py
#Thanks.
#from https://github.com/yuma-m/raspi_lcd_acm1602ni
from character_table import INITIALIZE_CODES, LINEBREAK_CODE, CHAR_TABLE

print(CHAR_TABLE["h"])
print(CHAR_TABLE["カ"])
print(CHAR_TABLE["あ"])





tests/classinclass.py
#!/user/bin/env python3

class A:
	AA = 5

	class B():
		BB = 50
		def __init__(self):
			pass
		def print(self,num):
			A.print(num)

	def __init__(self):
		self.b = self.B()

	@classmethod
	def print(self,num):
		print(self.AA,num)


if __name__ == '__main__':
	a1 = A()
	a2 = Qa.b.print(a.b.BB)


tests/character_table.py
../character_table.py

yoko_scroll.py
import time
import smbus
from time import sleep
from acm1602 import ACM1602
import threading
import subprocess

a = ACM1602()
a.cls()
a.speed(0.01)
a.sendMessage("コンニチハ",1)
a.pos[1] = 20 #カーソルの位置
a.pos[2] = 20 #カーソルの位置
a.stop = False #Trueでthread終了


messages1 = [" "]
messages2 = [" "]

class MyThread1(threading.Thread):
	def __init__(self):
		threading.Thread.__init__(self)

	def run(self):

		while True:
			a.screenHidari(1)
			a.pos[0] += 1
			time.sleep(0.1)
			if a.stop: break

class MyThread2(threading.Thread):
	def __init__(self):
		threading.Thread.__init__(self)

	def run(self):
		i=0	
		while True:
			if a.stop: break
			length = len(messages1)
			begin = a.pos[0]
			i += 1	
			bytes = a.send2Messages(messages1[i%length],messages2[i%length])
			bytes += a.send2Messages(" "," ")

			if i > 2*length:
				a.stop = True
				break

			while a.pos[0] != begin + bytes:
				time.sleep(0.05)



class MyThread3(threading.Thread):
	def __init__(self):
		threading.Thread.__init__(self)

	def run(self):
		while True:
			if a.stop: break
			nickname = "apache2"
			name = nickname
			check_service("NTP","ntp")
			check_service("APACHE2","apache2")
			check_service("WMAX31856","websocket-client-max31856-9801")
			check_service("sys_kama","sys_kama")
			check_usbmemory("USBメモリ","TOSHIBA")
			check_cputemp("CPUオンド")
			check_cpuvolt("CPUデンアツ")
			time.sleep(5)
			
def check_service(nickname,name):
	cmd = ("systemctl status {}".format(name))
	tmp = subprocess.Popen(cmd,stdout=subprocess.PIPE,shell=True).stdout.readlines()
	aa=""
	for ele in tmp:
		aa += ele.decode('utf-8')
	import re
	try:
		status = re.search("(?<=Active:\s)[^\s]+",aa).group(0)
	except AttributeError:
		status = "None"
	try:
		n = messages1.index(nickname)
		messages2[n] = status
	except ValueError:
		messages1.append(nickname)
		messages2.append(status)

def check_usbmemory(nickname,name):
	cmd = ("mount -l | grep {}".format(name))
	tmp = subprocess.Popen(cmd,stdout=subprocess.PIPE,shell=True).stdout.readlines()
	aa=""
	for ele in tmp:
		aa += ele.decode('utf-8')
	import re
	if aa:
		status = "OK"
	else:
		status = "BAD"

	try:
		n = messages1.index(nickname)
		messages2[n] = status
	except ValueError:
		messages1.append(nickname)
		messages2.append(status)

def check_cputemp(nickname):
	cmd = ("vcgencmd measure_temp")
	tmp = subprocess.Popen(cmd,stdout=subprocess.PIPE,shell=True).stdout.readlines()
	aa=""
	for ele in tmp:
		aa += ele.decode('utf-8')
	import re
	try:
		temp=re.search("(?<=temp=)[^\s]+",aa).group(0)
	except AttributeError:
		temp = "BAD"
	try:
		n = messages1.index(nickname)
		messages2[n] = temp
	except ValueError:
		messages1.append(nickname)
		messages2.append(temp)

def check_cpuvolt(nickname):
	cmd = ("vcgencmd measure_volts")
	tmp = subprocess.Popen(cmd,stdout=subprocess.PIPE,shell=True).stdout.readlines()
	aa=""
	for ele in tmp:
		aa += ele.decode('utf-8')
	import re
	try:
		temp=re.search("(?<=volt=)[^\s]+",aa).group(0)
	except AttributeError:
		temp = "BAD"
	try:
		n = messages1.index(nickname)
		messages2[n] = temp
	except ValueError:
		messages1.append(nickname)
		messages2.append(temp)


thread1 = MyThread1()
thread2 = MyThread2()
thread3 = MyThread3()
thread1.start()
thread2.start()
thread3.start()
thread1.join()
thread2.join()
thread3.join()







character_table.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

'''
The MIT License (MIT)

Copyright (c) 2016 Yuma.M

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''


INITIALIZE_CODES = [
    0x01,  # clear display
    0x38,  # function set: 8bit, 2lines, 5x8
    0x0c,  # display on, no cursor
    0x06,  # カーソルの動き
]

LINEBREAK_CODE = [0xc0]

CHAR_TABLE = {
    u'!': [0x21],
    u'"': [0x22],
    u'#': [0x23],
    u'$': [0x24],
    u'%': [0x25],
    u'&': [0x26],
    u"'": [0x27],
    u'(': [0x28],
    u')': [0x29],
    u'*': [0x2a],
    u'+': [0x2b],
    u',': [0x2c],
    u'-': [0x2d],
    u'.': [0x2e],
    u'/': [0x2f],

    u'0': [0x30],
    u'1': [0x31],
    u'2': [0x32],
    u'3': [0x33],
    u'4': [0x34],
    u'5': [0x35],
    u'6': [0x36],
    u'7': [0x37],
    u'8': [0x38],
    u'9': [0x39],
    u':': [0x3a],
    u';': [0x3b],
    u'<': [0x3c],
    u'=': [0x3d],
    u'>': [0x3e],
    u'?': [0x3f],

    u'@': [0x40],
    u'A': [0x41],
    u'B': [0x42],
    u'C': [0x43],
    u'D': [0x44],
    u'E': [0x45],
    u'F': [0x46],
    u'G': [0x47],
    u'H': [0x48],
    u'I': [0x49],
    u'J': [0x4a],
    u'K': [0x4b],
    u'L': [0x4c],
    u'M': [0x4d],
    u'N': [0x4e],
    u'O': [0x4f],

    u'P': [0x50],
    u'Q': [0x51],
    u'R': [0x52],
    u'S': [0x53],
    u'T': [0x54],
    u'U': [0x55],
    u'V': [0x56],
    u'W': [0x57],
    u'X': [0x58],
    u'Y': [0x59],
    u'Z': [0x5a],
    u'[[': [0x5b],
    u'¥': [0x5c],
    u']]': [0x5d],
    u'^': [0x5e],
    u'_': [0x5f],

    u'`': [0x60],
    u'a': [0x61],
    u'b': [0x62],
    u'c': [0x63],
    u'd': [0x64],
    u'e': [0x65],
    u'f': [0x66],
    u'g': [0x67],
    u'h': [0x68],
    u'i': [0x69],
    u'j': [0x6a],
    u'k': [0x6b],
    u'l': [0x6c],
    u'm': [0x6d],
    u'n': [0x6e],
    u'o': [0x6f],

    u'p': [0x70],
    u'q': [0x71],
    u'r': [0x72],
    u's': [0x73],
    u't': [0x74],
    u'u': [0x75],
    u'v': [0x76],
    u'w': [0x77],
    u'x': [0x78],
    u'y': [0x79],
    u'z': [0x7a],
    u'(': [0x7b],
    u'|': [0x7c],
    u')': [0x7d],
    u'→': [0x7e],
    u'←': [0x7f],

    u'。': [0xa1],
    u'「': [0xa2],
    u'」': [0xa3],
    u'、': [0xa4],
    u'・': [0xa5],
    u'ヲ': [0xa6],
    u"ァ": [0xa7],
    u'ィ': [0xa8],
    u'ゥ': [0xa9],
    u'ェ': [0xaa],
    u'ォ': [0xab],
    u'ャ': [0xac],
    u'ュ': [0xad],
    u'ョ': [0xae],
    u'ッ': [0xaf],

    u'ー': [0xb0],
    u'ア': [0xb1],
    u'イ': [0xb2],
    u'ウ': [0xb3],
    u'エ': [0xb4],
    u'オ': [0xb5],
    u'カ': [0xb6],
    u'キ': [0xb7],
    u'ク': [0xb8],
    u'ケ': [0xb9],
    u'コ': [0xba],
    u'サ': [0xbb],
    u'シ': [0xbc],
    u'ス': [0xbd],
    u'セ': [0xbe],
    u'ソ': [0xbf],

    u'タ': [0xc0],
    u'チ': [0xc1],
    u'ツ': [0xc2],
    u'テ': [0xc3],
    u'ト': [0xc4],
    u'ナ': [0xc5],
    u'ニ': [0xc6],
    u'ヌ': [0xc7],
    u'ネ': [0xc8],
    u'ノ': [0xc9],
    u'ハ': [0xca],
    u'ヒ': [0xcb],
    u'フ': [0xcc],
    u'ヘ': [0xcd],
    u'ホ': [0xce],
    u'マ': [0xcf],

    u'ミ': [0xd0],
    u'ム': [0xd1],
    u'メ': [0xd2],
    u'モ': [0xd3],
    u'ヤ': [0xd4],
    u'ユ': [0xd5],
    u'ヨ': [0xd6],
    u'ラ': [0xd7],
    u'リ': [0xd8],
    u'ル': [0xd9],
    u'レ': [0xda],
    u'ロ': [0xdb],
    u'ワ': [0xdc],
    u'ン': [0xdd],
    u'゛': [0xde],
    u'゜': [0xdf],

    u'α': [0xe0],
    u'β': [0xe2],
    u'ε': [0xe3],
    u'μ': [0xe4],
    u'δ': [0xe5],
    u'ρ': [0xe6],
    u'√': [0xe8],

    u'∞': [0xf3],
    u'Ω': [0xf4],
    u'Σ': [0xf6],
    u'π': [0xf7],
    u'千': [0xfa],
    u'万': [0xfb],
    u'円': [0xfc],
    u'÷': [0xfd],
    u' ': [0xfe],
    u'　': [0xfe],
    u'■': [0xff],

    u'ガ': [0xb6, 0xde],
    u'ギ': [0xb7, 0xde],
    u'グ': [0xb8, 0xde],
    u'ゲ': [0xb9, 0xde],
    u'ゴ': [0xba, 0xde],
    u'ザ': [0xbb, 0xde],
    u'ジ': [0xbc, 0xde],
    u'ズ': [0xbd, 0xde],
    u'ゼ': [0xbe, 0xde],
    u'ゾ': [0xbf, 0xde],
    u'ダ': [0xc0, 0xde],
    u'ヂ': [0xc1, 0xde],
    u'ヅ': [0xc2, 0xde],
    u'デ': [0xc3, 0xde],
    u'ド': [0xc4, 0xde],
    u'バ': [0xca, 0xde],
    u'ビ': [0xcb, 0xde],
    u'ブ': [0xcc, 0xde],
    u'ベ': [0xcd, 0xde],
    u'ボ': [0xce, 0xde],
    u'パ': [0xca, 0xdf],
    u'ピ': [0xcb, 0xdf],
    u'プ': [0xcc, 0xdf],
    u'ペ': [0xcd, 0xdf],
    u'ポ': [0xce, 0xdf],
}


acm1602.py
import time
import smbus
from time import sleep

#Thanks.
#from https://github.com/yuma-m/raspi_lcd_acm1602ni
from character_table import INITIALIZE_CODES, LINEBREAK_CODE, CHAR_TABLE

class ACM1602:
	BUS_ADDR = 1
	SLAVE_ADDR = 0x50
	CONTROL_SET = 0x00
	CONTROL_WRITE = 0x80

	#default
	bCD = 1
	bRH = 2
	bEMS = 6
	bDOOC = 12
	bFS = 0x38
	bCDS = 0x1C

	def __init__(self):
		self.pos = [0,0,0]
		self.lcd = smbus.SMBus(self.BUS_ADDR)
		self.sleeptime = 0.01

		#make instance
		self.entryModeSet       = self.EntryModeSet()
		self.displayOnOffControl= self.DisplayOnOffControl()
		self.cursorDisplayShift = self.CursorDisplayShift()
		self.functionSet        = self.FunctionSet()

		#override Base class' property 'parent'
		self.entryModeSet.parent        = self
		self.displayOnOffControl.parent = self
		self.cursorDisplayShift.parent  = self
		self.functionSet.parent         = self

		#initialize
		self.functionSet.set(self.FunctionSet.DATA8_LINE2_YDOTS10)
		self.clearDisplay()
		self.entryModeSet.set(self.EntryModeSet.MOJIWOUTTARA_CURSOR_MIGI)
		self.displayOnOffControl.set(self.DisplayOnOffControl.DISPON_CURSORON_BLINKON)

	def version(self):
		return '1.0'

	class Base:
		def __init__(self):
			self.parent = null #interface like Java This must be overrided when this class is inherited.
		def set(self,num):
			self.parent._set(num)
		def dict(self):
			for ele in self.__dict__:
				print(ele)

	def clearDisplay(self):
			self._set(1)
			self.pos[1] = 0
			self.pos[2] = 0

	def returnHome(self):
			self._set(2)


	class EntryModeSet(Base):
		MOJIWOUTTARA_CURSOR_MIGI   = 0x6
		MOJIWOUTTARA_CURSOR_HIDARI = 0x4
		MOJIWOUTTARA_GAMEN_MIGI     = 0x5
		MOJIWOUTTARA_GAMEN_HIDARI   = 0x7
		def __init__(self):
			pass
		
	class DisplayOnOffControl(Base):
		DISPON_CURSORON_BLINKON    = 0xF
		DISPON_CURSORON_BLINKOFF   = 0xE
		DISPON_CURSOROFF_BLINKON   = 0xD
		DISPON_CURSOROFF_BLINKOFF  = 0xC
		DISPOFF_CURSORON_BLINKON   = 0xB
		DISPOFF_CURSORON_BLINKOFF  = 0xA
		DISPOFF_CURSOROFF_BLINKON  = 0x9
		DISPOFF_CURSOROFF_BLINKOFF = 0x8
		def __init__(self):
			pass
	
	class CursorDisplayShift(Base):
		SCREEN_HIDARI = 0x18
		SCREEN_MIGI = 0x1C
		CURSOR_HIDARI = 0x10
		CURSOR_MIGI =0x14
		def __init__(self):
			pass

	class FunctionSet(Base):
		DATA8_LINE2_YDOTS8  = 0x3C
		DATA8_LINE2_YDOTS10 = 0x38
		DATA8_LINE1_YDOTS8  = 0x34
		DATA8_LINE1_YDOTS10 = 0x30
		DATA4_LINE2_YDOTS8  = 0x2C
		DATA4_LINE2_YDOTS10 = 0x28
		DATA4_LINE1_YDOTS8  = 0x24
		DATA4_LINE1_YDOTS10 = 0x20
		def __init__(self):
			pass
	
	def setCGRAMAddress(self,xx):
		xx = xx | 0x40
		self._set(xx)
	
#	def setDDRAMAddress(self,gyou,retsu):
#		if gyou < 1:gyou = 1
#		if gyou > 2:gyou = 2
#		if retsu < 1:retsu = 1
#		if retsu > 16:retsu = 16
#		addr = (gyou - 1) * 0x40 + retsu -1;	
#		addr = addr | 0x80
#		self._set(addr)

	def setDDRAMAddress(self,xx):
		xx = xx | 0x80
		self._set(xx)


	def sendMessage(self,strs,nLine):
		nByte = 0
		for ii in range(len(strs)):
			nByte += self._put1(strs[ii],nLine)
		return nByte

	def send2Messages(self,str1,str2):
		bytes1 = self.sendMessage(str1,1)
		bytes2 = self.sendMessage(str2,2)

		if bytes1 > bytes2:
			self.sendMessage(' ' * (bytes1 -bytes2),2)
		elif bytes2 > bytes1:
			self.sendMessage(' ' * (bytes2 -bytes1),1)

		return bytes1 if bytes1 > bytes2 else bytes2


	def _set(self,num):
		if self.lcd is not None:
			self.lcd.write_byte_data(self.SLAVE_ADDR,self.CONTROL_SET,num)
			time.sleep(0.01)
		else:
			self._message_not_open(self)#クラスメソッドの中からインスタンスメソッドを呼ぶときにはselfを渡す。
	def _put1(self,str,nLine):

		#nLine > 2 or nLine < 1 then エラー発生


		if self.lcd is not None:
			if str == " " or str == "　":
				sp = 0.01
			else:
				sp = self.sleeptime
			try:
				bytes = CHAR_TABLE[str]
			except KeyError:
				bytes = CHAR_TABLE["X"]
			if len(bytes) == 2:
				sp = 0.01
			for byte in bytes:
				if nLine == 1:
					self.setDDRAMAddress(self.pos[nLine])
				else:
					self.setDDRAMAddress(0x40 + self.pos[nLine])

				if self.pos[nLine] > 39:
					if nLine == 1:
						self.setDDRAMAddress(0)
					else:
						self.setDDRAMAddress(0x40)
					self.pos[nLine] = 0
				self.lcd.write_byte_data(self.SLAVE_ADDR,self.CONTROL_WRITE,int(byte))
				self.pos[nLine] += 1
				#time.sleep(sp)
				sp = self.sleeptime
		else:
			self._message_not_open()
		return len(bytes)


	def _message_not_open(self):
		print(self)
		print("self.lcd has been None yet.")

	def speed(self,sp):
		if sp < 0.01:sp = 0.01
		self.sleeptime = sp
	#sugars

	def cls(self):
		self.clearDisplay()
	def cls1(self):
		self.line1()
		self.sendMessage("                 ")
	def cls2(self):
		self.line2()
		self.sendMessage("                 ")
	def line1(self):
		self.setDDRAMAddress(0)
	def line2(self):
		self.setDDRAMAddress(0x40)
		

	def screenMigi(self,nn):
		for i in range(nn):
			self.cursorDisplayShift.set(self.CursorDisplayShift.SCREEN_MIGI)
	def screenHidari(self,nn):
		for i in range(nn):
			self.cursorDisplayShift.set(self.CursorDisplayShift.SCREEN_HIDARI)
	def cursorHidari(self,nn):
		for i in range(nn):
			self.cursorDisplayShift.set(self.CursorDisplayShift.CURSOR_HIDARI)
	def cursorMigi(self,nn):
		for i in range(nn):
			self.cursorDisplayShift.set(self.CursorDisplayShift.CURSOR_MIGI)
			



if __name__ == '__main__':

	a = ACM1602()

	print(a.pos)

	#画面に表示する
	if True:
		a.cls()
		a.speed(0.01)
		a.sendMessage("コンニチハ",1)
	#	a.line2()
	#	a.sendMessage("キョウハイイテンキデスネ",1)
	#push命令を作る
	#行を指定して文字列をpushすると、現在の表示がだんだんと左にずれて、pushした文字列が右側から現れるという機能
	if False:
		a.cls()
		a.entryModeSet.set(a.EntryModeSet.MOJIWOUTTARA_CURSOR_MIGI)
		#a.sendMessage('     o                             ')
		#a.entryModeSet.set(a.EntryModeSet.MOJIWOUTTARA_GAMEN_HIDARI)
		i=0
		for n in range(1,41):
			a.sendMessage(str(n%10))	
			i += 1
			print(i)
			if i > 16:
				a.screenHidari(1)
				pass
		a.setDDRAMAddress(0x40 + 39 )

		a.sendMessage("A")	
	if True:
		import threading

		class MyThread1(threading.Thread):
			def __init__(self):
				threading.Thread.__init__(self)

			def run(self):

				while True:
					a.screenHidari(1)
					a.pos[0] += 1
					time.sleep(0.1)

		class MyThread2(threading.Thread):
			def __init__(self):
				threading.Thread.__init__(self)

			def run(self):
				i=0	
				stringss1 = ["コンヤノテンキ","グンマケン","サイタマケン","トチギケン"]
				stringss2 = ["","ハレ","ハレ","クモリ"]
				while True:
					begin = a.pos[0]
					i += 1	
					bytes = a.send2Messages(stringss1[i%3],stringss2[i%3])
					bytes += a.send2Messages(" "," ")

					while a.pos[0] != begin + bytes:
						time.sleep(0.05)

		a.pos[1] = 20
		a.pos[2] = 20
		thread1 = MyThread1()
		thread2 = MyThread2()
		thread1.start()
		thread2.start()
		thread1.join()
		thread2.join()





